### PR TITLE
Fix CR about evaluation order in To_cmm_primitive etc.

### PR DESCRIPTION
**Pull request by mshinwell**
This changes the semantics of the expert "inline effects in Cmm" flag.  This flag is off by default.  After this PR the "off" state has a different meaning: inlining of effects is still permitted into `Make_block` and `Make_array` primitives.  The comments explain what is going on.

### CODE REVIEW

**Comment by mshinwell** [comment]
sorry this still isn't quite right, will remove "draft" when fixed

**Comment by mshinwell** [comment]
Hopefully this is better

File: middle_end/flambda2/to_cmm/to_cmm_env.ml [hunk 1]
```diff
@@ -326,6 +326,10 @@ let inline_variable env var =
       if not (Variable.equal var var_from_stage)
       then will_not_inline_var env var
       else if binding.may_inline
+              &&
+              match consider_inlining_effectful_expressions with
+              | Some consider -> consider
+              | None -> Flambda_features.Expert.inline_effects_in_cmm ()
```
Comment (Gbury) [inline middle_end/flambda2/to_cmm/to_cmm_env.ml:332]: maybe that match could be lifted out of the condition for the if-then-else

File: middle_end/flambda2/to_cmm/to_cmm_primitive.ml [hunk 2]
```diff
@@ -692,31 +692,68 @@ let variadic_primitive _env dbg f args =
...
+       [y] on the most recent stage since it was the most recent binding. The
+       [Simple] for [y] will be translated first by the code below, meaning
+       inlining is permitted (since [y] is on the most recent stage), producing
+       Make_block [effect-y; y]. Then the [Simple] for [x] will be translated,
+       producing the desired output Make_block [effect-y; effect-x]. The backend
+       will compile this to run effect-x before effect-y by virtue of
+       right-to-left evaluation order. This therefore matches the original code.
+
+       (Same for the binary and ternary cases above, although it usually isn't
+       relevant, as inline-effects-in-cmm is defaulted off.) *)
```
Comment (Gbury) [inline middle_end/flambda2/to_cmm/to_cmm_primitive.ml:754]: Note that this comment also applies to the other primitive translations (where translate primitive arguments left-to-right)

**Review by Gbury — APPROVED** [review]
Looks good and should fix the potential problems with order of evaluation. In any case, the same parts of code will have to be modified again for the upcoming unboxing in to_cmm work.

File: middle_end/flambda2/to_cmm/to_cmm_shared.mli [hunk 3]
```diff
@@ -37,14 +37,17 @@ val symbol_from_linkage_name :
...
 
+(** This does not inline effectful expressions. *)
 val name :
   To_cmm_env.t ->
   Name.t ->
   Cmm.expression * To_cmm_env.t * Effects_and_coeffects.t
 
 val const : dbg:Debuginfo.t -> Reg_width_const.t -> Cmm.expression
 
+(** [consider_inlining_effectful_expressions] defaults to [false]. *)
```
Comment (lthls) [inline middle_end/flambda2/to_cmm/to_cmm_shared.mli:48]: The default behaviour is to use `Flambda_features.Expert.inline_effects_in_cmm`, which itself defaults to false if no flag is given. Maybe the comment could reflect that in some way ?

**Review by lthls — APPROVED** [review]

File: middle_end/flambda2/to_cmm/to_cmm_primitive.ml (hunk 2 above)
Comment (mshinwell) [reply middle_end/flambda2/to_cmm/to_cmm_primitive.ml:754]: I don't understand this comment, can you elaborate?

File: middle_end/flambda2/to_cmm/to_cmm_env.ml (hunk 1 above)
Comment (mshinwell) [reply middle_end/flambda2/to_cmm/to_cmm_env.ml:332]: Fixed

File: middle_end/flambda2/to_cmm/to_cmm_shared.mli (hunk 3 above)
Comment (mshinwell) [reply middle_end/flambda2/to_cmm/to_cmm_shared.mli:48]: Fixed

File: middle_end/flambda2/to_cmm/to_cmm_shared.mli [hunk 4]
```diff
@@ -53,7 +56,10 @@ val simple :
 val simple_static :
   Simple.t -> [`Data of Cmm.data_item list | `Var of Variable.t]
 
+(** This function translates the [Simple] at the head of the list first.
+    [consider_inlining_effectful_expressions] defaults to [false]. *)
```
[the remark this answers is not in the document]
Comment (mshinwell) [reply middle_end/flambda2/to_cmm/to_cmm_shared.mli:60]: Fixed

**Pull request merged**